### PR TITLE
enable UNMAP when LUN is thin provisioned

### PR DESCRIPTION
**Pull request by chessman**
- support Block Limits VPD page (0xB0)
 - add UNMAP to REPORT SUPPORTED OPERATION CODES
 - READ CAPACITY(16): set LBPME when Thin provisioning is enabled
 - move Thinprovisioning and BlockShift to config
 - add Unmap to BackingStore

### CODE REVIEW

**Comment by chessman** [comment]
That's a work in progress. I still need to parse UNMAP block descriptors.

**Comment by carmark** [comment]
@chessman Awesome

**Comment by orzhang** [comment]
@chessman Awesome. :-)  We are looking forward to the implementation of unmap operation in backing store. BTW Which kind of project are you working for?

**Comment by chessman** [comment]
I've added Unmap to backing store, could you look at the PR?

@orzhang I'm implementing an iSCSI interface to commercial cloud storage. This storage is used as a backing store. It should be thin provisioned because it exposes more space than it actually has and it needs UNMAP to cleanup unused blocks.

The next thing I want to address is caching. SYNCHRONIZE CACHE command doesn't reach a target now. In logs I see:
```
[944080.680374] sd 6:0:0:0: [sdc] No Caching mode page found
[944080.680378] sd 6:0:0:0: [sdc] Assuming drive cache: write through
```
It looks like there is a problem with MODE SENSE Caching page.

**Comment by chessman** [comment]
I added Logical Block Provisioning VPD page support as well to satisfy libiscsi tests.

**Comment by chessman** [comment]
@carmark Hello! I removed WriteAtomic16 tests from travis. The reason they worked prior to this PR is that BlockLimits VPD page was not implemented, the tests were actually skipped. I think, now I finished with this PR.

File: pkg/config/config.go [hunk 1]
```diff
@@ -101,9 +101,11 @@ var (
 )
 
 type BackendStorage struct {
-	DeviceID uint64 `json:"deviceID"`
-	Path     string `json:"path"`
-	Online   bool   `json:"online"`
+	DeviceID         uint64 `json:"deviceID"`
+	Path             string `json:"path"`
+	Online           bool   `json:"online"`
+	Thinprovisioning bool   `json:"thinprovisioning"`
```
Comment (carmark) [inline pkg/config/config.go]:
```
ThinProvisioning bool `json:"thinProvisioning"`
```
?

File: pkg/config/config.go [hunk 2]
```diff
@@ -101,9 +101,11 @@ var (
...
+	BlockShift       uint   `json:"blockShift"`
```
Comment (carmark) [inline pkg/config/config.go]: could u please add these two new fields into [examples/config.json](https://github.com/gostor/gotgt/blob/master/examples/config.json#L4)?

File: pkg/scsi/lun.go [hunk 3]
```diff
@@ -43,8 +43,8 @@ func NewSCSILu(device_uuid uint64, path string, online bool) (*api.SCSILu, error
 		PerformCommand: luPerformCommand,
 		DeviceProtocol: sbc,
 		Storage:        backing,
-		BlockShift:     api.DefaultBlockShift,
-		UUID:           device_uuid,
+		BlockShift:     bs.BlockShift,
```
Comment (carmark) [inline pkg/scsi/lun.go]: Shall we set `BlockShift` as `api.DefaultBlockShift` if the user does not set `BlockShift`?

Comment (chessman) [reply pkg/scsi/lun.go]: A fallback to the default value happens in sbc.go:96. Do you want to do it here as well?

File: pkg/config/config.go (hunk 2 above)
Comment (chessman) [reply pkg/config/config.go]: Done

File: pkg/config/config.go (hunk 1 above)
Comment (chessman) [reply pkg/config/config.go]: Done

**Comment by carmark** [comment]
LGTM and thanks.

**Pull request merged**